### ISSUE TITLE
Page replaceState issue with Proxy Vue 3

### ISSUE THREAD

**Issue by rwdevguy** [inertiajs/inertia#309]
Current receiving an issue with trying to use Laravel Jetstream's Inertia Form:
```
Uncaught DOMException: Failed to execute 'replaceState' on 'History': [object Object] could not be cloned.
    at Proxy.replaceState (http://localhost:3000/js/admin.js:24884:39)
    at Proxy.saveScrollPositions (http://localhost:3000/js/admin.js:24697:14)
    at Proxy.visit (http://localhost:3000/js/admin.js:24798:73)
    at Proxy.put (http://localhost:3000/js/admin.js:24923:73)
    at Proxy.submit (http://localhost:3000/js/admin.js:51142:47)
    at Proxy.put (http://localhost:3000/js/admin.js:51100:25)
    at Proxy.onSubmit (http://localhost:3000/js/admin.js:42751:17)
    at callWithErrorHandling (http://localhost:3000/js/admin.js:31936:22)
    at callWithAsyncErrorHandling (http://localhost:3000/js/admin.js:31945:21)
    at emit (http://localhost:3000/js/admin.js:32390:9)
```

This seems to be because the `page` argument that is getting passed to `replaceState` is actually a proxy since the Jetstream Inertia Form instance is part of the component's data object.

I did see that [#297](https://github.com/inertiajs/inertia/pull/297) had been merged to fix similar errors but this one still persists.

Currently using :

@inertiajs/inertia **0.6.1**
@inertiajs/inertia-vue3 **0.2.1**

(the issue existed on previous version as well).

A fix that seems to solve the problem is:

```
  pushState(page) {
    this.page = page
    const clone = JSON.parse(JSON.stringify(page))
    window.history.pushState(clone, '', page.url)
  },

  replaceState(page) {
    this.page = page
    const clone = JSON.parse(JSON.stringify(page))
    window.history.replaceState(clone, '', page.url)
  },
```

Happy to make a PR for this but completely understand if you feel the issue is out of scope of since it mainly relates to a different package.

**Comment by claudiodekker** [comment]
Hi,

I'd be happy to take a look at this. What version of `jetstream-js` do you have installed?
Also, could you show me the violating piece of code, so we can use that to replicate it?

Thanks!

**Comment by reinink** [comment]
Maybe @m1guelpf can help as well, because he apparently ran into the [same issue](https://github.com/inertiajs/inertia/pull/297#issuecomment-718832888).

**Comment by m1guelpf** [comment]
@rwdevguy Could you upgrade `jestream-js` to `v1.1.0` and try again?

**Comment by rwdevguy** [comment]
@m1guelpf I've updated to `v1.1.0`, however, I'm using the package globally so it uses the preexisting install method rather than the `useForm()` hook which you introduced.

I can confirm that using `shallowReactive` around `this.$inertia.form({...})` does seem to fix the issue.

@claudiodekker Here is an example:
**app.js**
```
import {createApp, h} from "vue";
import {app, plugin} from "@inertiajs/inertia-vue3";
import {InertiaForm} from "laravel-jetstream";

const el = document.getElementById("app");

createApp({
    render: () =>
        h(app, {
            initialPage: JSON.parse(el.dataset.page),
            resolveComponent: name => require(`@/Pages/${name}`).default,
        }),
})
    .use(plugin)
    .use(InertiaForm)
    .mount(el);

```

**Test.vue**
```
<template>
    <div>
        <button @click="submitWithInertia">
            Submit with Inertia
        </button>
        <button @click="submitWithLaravelJetstream">
            Submit with Laravel Jetstream Inertia Form
        </button>
    </div>
</template>

<script>
export default {
    data() {
        return {
            form: this.$inertia.form({
                test: "test",
            }),
        };
    },

    methods: {
        submitWithInertia() {
            this.$inertia.post("/some-endpoint", this.form.data());
        },
        submitWithLaravelJetstream() {
            this.form.post("/some-endpoint");
        },
    },
};
</script>

```

**Comment by m1guelpf** [comment]
@rwdevguy This is definitely related to `laravel-jetstream` then, I'm guessing it'd be best to close here and open there.

**Comment by claudiodekker** [comment]
@m1guelpf Yeah, I already figured from the first message that this was the case, but it really isn't a huge issue to keep the issue here and to have us look into it, as the plan is for `jetstream-js` to be consumed into Inertia entirely before 1.0.0 hits anyway 👍

**Comment by reinink** [comment]
Okay, so I'm pretty sure the issue here is local state caching (the "remember" feature). We actually ran into the same problem recently while adding the Jetstream form helper to the Vue 2 and Vue 3 adapters. The issue is that the Jetstream form object cannot be serialized when being saved to the history state (via the remember feature). That's because it's a class, with functions, and not just plain data.

We solved this in our new form helper using new `serialize` and `unserialize` methods, that Inertia will automatically check for and call when saving the remember data to history state.

https://github.com/inertiajs/inertia/blob/17e6c711d91d7ab33bd3557b032a8ddcac6b124b/packages/inertia-vue/src/form.js#L59-L68

Meaning, this is a Jetstream issue, but one that we've solved in the latest version of Inertia. All these changes are being brought to Jetstream 2.0. 👍

**Comment by petethewizard** [comment]
I am having this issue with inertia:

"@inertiajs/inertia": "^0.10.0",
"@inertiajs/inertia-vue3": "^0.5.1",
"@inertiajs/progress": "^0.2.6",

and

vue: 3.0.5

and

"laravel/jetstream": "^2.3",

I've zipped the component in which it happens and attached as copying and pasting it here is not displaying it properly.

[Datatable.zip](https://github.com/inertiajs/inertia/files/7456971/Datatable.zip)

**Comment by samuelsennev** [comment]
Hey guys! 

I have this `laravel-inertia` application with a presence channel chat and I've been facing this issue as well (still didn't fix it).

**Settings**
`inertiajs/inertia-laravel: ^0.4.3 `
`@inertiajs/inertia: ^0.10.0`
`@inertiajs/inertia-vue3: ^0.5.1`

**What is happening**

With the presence channel set, when I connect to it and then leave it, everything works perfectly. 

The error occurs when, in a second scenario, when I try to leave the channel after sending a message.

As I said, when I dont send a message, everything works fine, I can connect to it and leave it, without triggering this exception. 

There is no error related with broadcast event, as well with the message sent. The exception is triggered when I try to leave the channel after, and only after, i send a message.

**Comment by heleonprime** [comment]
Hello. Is there anybody? I got the same issue in my code, but not with request sending functions. 
I'm trying to change page state when changing table ordering.
I've tried like this
```
this.$inertia.pushState(this.$page)
```
Then i found that inside inertia's pushState method there's also `this.page` update, 
And then i've tried to do it in my component on click:
```
this.$page.url = this.pageBaseUrl + this.getUrlParams(pg);
this.$inertia.page = {...this.$page};
history.replaceState({}, '', this.$page.url);
```
And it worked, but until i've tried to scroll the page
When i scroll the page i got 
![image](https://user-images.githubusercontent.com/7357817/179274326-cc3b66c8-852a-4323-9a2f-92092312afdd.png)
And then i found that the problen is in `this.$inertia.page = {...this.$page};` string, if i comment it the error's dissaper

How can i set the page properly. When i'm trying to set the state without setting the page - the url state refreshes to prevoius state on scroll. Why the scroll changes the state? 

Here's the part of my code, there's a many logs, i'll remove'em later, but now it halped me to find this problem place
```
    l(...messages)
    {
      if(this.debug)
      {
        console.log(...messages);
      }
    },
    reset() {
      this.form = mapValues(this.form, () => null)
    },
    toggleOrder(name)
    {
      this.l('getUrlParams');
      this.l('     field name', name);
      this.l('     field name now', this.form.order[name]);
      if(this.form.order[name])
      {
        if(this.form.order[name] === 'asc')
          this.form.order[name] = 'desc'.toLowerCase();
        else if(this.form.order[name] === 'desc')
          delete this.form.order[name];
      }
      else
        this.form.order[name] = 'asc'.toLowerCase();
      this.l('     field name after', this.form.order[name]);
      this.l('     inithializing getList with page = 1');
      this.getList(1)
    },
    addHistory(pg)
    {
      this.l('getUrlParams');
      this.l('     pg:', pg);
      this.l('     this.getUrlParams(pg)', this.getUrlParams(pg));
      this.l('pushState:', this.$inertia.pushState);
      this.$page.url = this.pageBaseUrl + this.getUrlParams(pg);
      this.$inertia.page = {...this.$page};
      this.l(this.$page.url);
      this.l(this.$page.target);
      this.l({...this.$page});
      this.l(JSON.parse(JSON.stringify(this.$page)));
      //history.replaceState({}, '', this.$page.url)

      //this.$inertia.pushState(this.$page) /*{} ,null, this.pageBaseUrl + this.getUrlParams(pg)*/
      this.l('$inertia:', this.$inertia);
    },
    getUrlParams(pg){
      this.l('getUrlParams');
      this.l('     pg: ', pg);
      this.l('!!!!!TRYING URLSearchParams():', new URLSearchParams(this.form).toString());
      let params = [];
      let orderKeys = Object.keys({...this.form.order});
      this.l('     orderKeys length: ', orderKeys.length);
      if(orderKeys.length)
        orderKeys.forEach((val) => {
          params.push(`order[${val}]=${this.form.order[val]}`);
        });
      if (this.form.search)  params.push(`search=${this.form.search}`);
      if (this.form.trashed) params.push(`trashed=${this.form.trashed}`);
      if (this.form.active)  params.push(`active=${this.form.active}`);
      if (pg) params.push(`page=${pg}`);
      this.l('     all params: ', params);
      this.l('     params length: ', params.length);
      this.l('     params string: ', params.length ? `?${params.join('&')}`: '');
      if (params.length) return `?${params.join('&')}`;
      return '';
    },
    getList(pg)
    {
      this.l('getList');
      this.l('     pg:', pg);
      this.l('     nulling last_page:', !pg || pg === 1);
      if(!pg || pg === 1) this.last_page = null;
      this.l('     loading:', this.loading);
      this.l('     last_page:', this.last_page);
      this.l('     current_page:', this.current_page);
      this.l('     can go on:', !this.loading && (!this.last_page || ( this.last_page && this.current_page < this.last_page)));
      if(!this.loading && (!this.last_page || ( this.last_page && this.current_page < this.last_page)))
      {
        this.loading = 1;
        this.l('     loading:', this.loading);
        this.form.page = pg;
        this.l('     form.page:', this.form.page);
        this.l('     sending request');
        this.$http.get('/api/parsers' + this.getUrlParams(pg),{
          withCredentials: true,
          //params: this.form,
        }).then((response) => {
          this.l('     got response', response);
          this.l('     rewriting pages list:', !pg || pg === 1);
          if(!pg || pg === 1) this.ps = response.data.parsers.data;
          else this.ps = this.ps.concat(response.data.parsers.data);

          this.current_page = response.data.parsers.current_page;
          this.l('     current_page:', this.current_page);
          this.last_page = response.data.parsers.last_page;
          this.l('     last_page:', this.last_page);
          this.psc = response.data.count;
          this.l('     pages count:', this.psc);
          this.loading = 0;
          this.l('     loading:', this.loading);
          this.l('     inithializing addHistory');
          this.addHistory(pg);
        });
      }
    },
    onScroll(event) {
      this.l('onScroll');
      this.lastPageIsVisible = useElementVisibility(this.lastPage).value;
      this.l('     lastPageIsVisible:', this.lastPageIsVisible);
      if(this.lastPageIsVisible)
      {
        this.l('     inithializing getList with page = ' + this.current_page + 1);
        this.getList(this.current_page + 1);
      }
    },
```
Here's what i got on scroll 
![image](https://user-images.githubusercontent.com/7357817/179276965-7656efee-0449-4d13-9bc2-857847672786.png)
